power_grid_model_c/power_grid_model/include/power_grid_model/component/load_gen.hpp

```diff
@@ -87,14 +87,26 @@ class LoadGen final : public std::conditional_t<is_gen, GenericGenerator, Generi
     }
 
     // update for load_gen
-    UpdateChange update(LoadGenUpdate<sym> const& update) {
-        assert(update.id == this->id());
-        this->set_status(update.status);
-        set_power(update.p_specified, update.q_specified);
+    UpdateChange update(LoadGenUpdate<sym> const& update_data) {
+        assert(update_data.id == this->id());
+        this->set_status(update_data.status);
+        set_power(update_data.p_specified, update_data.q_specified);
         // change load connection and/or value will not change topology or parameters
         return {false, false};
     }
 
+    LoadGenUpdate<sym> inverse(LoadGenUpdate<sym> update_data) const {
+        double const scalar = direction_ * base_power<sym>;
+
+        assert(update_data.id == this->id());
+
+        set_if_not_nan(update_data.status, static_cast<IntS>(this->status()));
+        set_if_not_nan(update_data.p_specified, real(s_specified_) * scalar);
+        set_if_not_nan(update_data.q_specified, imag(s_specified_) * scalar);
```

Comment:
I do not understand why this will work. You try to generate an inverse update by reading the current value in the component class. But the data members in the component class are not one-to-one mapping to the values in the update struct. In case of `LoadGen`, the values stored inside the class is p.u. value. But the value in the update struct is using `W` and `var` unit. You need to than back converts the values when you generate the inverse update. 

This is straightforward for `LoadGen`, but might be very complicated for other values.

Reply:
that's what the `scalar` is used for. that's the responsibility of the in-class implementation. I added explicit tests or every component to check that the initially provided value in the input data has the same value as the result data.

I personally also believe that it's the classes' own responsibility to ensure that the output matches closely enough that there is no value drift due to floating-point rounding errors, in which case both values would need to be stored. That would not introduce an unreasonably large memory overhead, because the components are no longer copied. I did not implement that yet, though. Do you think that I should implement that now?

Reply:
I don't think that's needed unless it is proven that certain operations hit numerical issues.

Reply:
well there's a potential drift for every scenario. if there's a batch of 10k scenarios, then it could become an issue

Reply:
We are not doing incremental stuff here. It only a multiplication. The drift is really minimum.

As for batch calculations, you always return the original value by `p_specified_org_ * scale`. That is always one drift, regardless of hoe many scenario you have.

Reply:
the drift may occur when doing `input(p_specified / scale) -> inverse(p_specified / scale * scale) -> update(p_specified / scale * scale / scale) -> inverse(p_specified / scale * scale / scale * scale) -> update(p_specified / scale * scale / scale * scale / scale) -> ...` but i guess indeed the difference is minimal